docs/framework/react/guides/custom-errors.md

```diff
@@ -0,0 +1,267 @@
+
+```tsx
+<form.Field
+  name="email"
+  validators={{
+    onChange: ({ value }) => {
+      if (!value.includes("@")) {
+        return {
+          message: "Invalid email format",
+          severity: "error",
+          code: 1001
+        };
+      }
+      return undefined;
+    }
+  }}
+/>
+```
+
+Display in UI:
+```tsx
+{typeof field.state.meta.errors[0] === "object" && (
+  <div className={`error ${field.state.meta.errors[0].severity}`}>
+    {field.state.meta.errors[0].message}
+    <small> (Code: {field.state.meta.errors[0].code})</small>
+  </div>
+)}
+```
+
+or
+```tsx
+{field.state.meta.errorMap.onChange?.message ? (
+    <p>{field.state.meta.errorMap.onBlur?.message}</p>
+  ) : null}
+```
+
+in the example above it depends on the event error you want to display either `onChange` `onBlur` or else, but of course with TypeScript support helping you out
+
+### Arrays
+
+Multiple error messages for a single field:
+
+```tsx
+<form.Field
+  name="password"
+  validators={{
+    onChange: ({ value }) => {
+      const errors = [];
+      if (value.length < 8) errors.push("Password too short");
+      if (!/[A-Z]/.test(value)) errors.push("Missing uppercase letter");
+      if (!/[0-9]/.test(value)) errors.push("Missing number");
+
+      return errors.length ? errors : undefined;
+    }
+  }}
+/>
+```
+
+Display in UI:
+```tsx
+{Array.isArray(field.state.meta.errors) && (
+  <ul className="error-list">
+    {field.state.meta.errors.map((err, i) => (
+      <li key={i}>{err}</li>
+    ))}
+  </ul>
+)}
+```
+
+## The `disableErrorFlat` Prop on Fields
+
+By default, TanStack Form flattens errors from all validation sources (onChange, onBlur, onSubmit) into a single `errors` array. The `disableErrorFlat` prop preserves the error sources:
+
+```tsx
+<form.Field
+  name="email"
+  disableErrorFlat
+  validators={{
+    onChange: ({ value }) => !value.includes("@") ? "Invalid email format" : undefined,
+    onBlur: ({ value }) => !value.endsWith(".com") ? "Only .com domains allowed" : undefined,
+    onSubmit: ({ value }) => value.length < 5 ? "Email too short" : undefined
+  }}
+/>
+```
+
+Without `disableErrorFlat`, all errors would be combined into `field.state.meta.errors`. With it, you can access errors by their source:
+
+```tsx
+{/* Access errors by validation source */}
+{field.state.meta.errorMap.onChange && (
+  <div className="real-time-error">{field.state.meta.errorMap.onChange}</div>
+)}
+
+{field.state.meta.errorMap.onBlur && (
+  <div className="blur-feedback">{field.state.meta.errorMap.onBlur}</div>
+)}
+
+{field.state.meta.errorMap.onSubmit && (
+  <div className="submit-error">{field.state.meta.errorMap.onSubmit}</div>
+)}
+```
+
+This is useful for:
+- Displaying different types of errors with different UI treatments
+- Prioritizing errors (e.g., showing submission errors more prominently)
+- Implementing progressive disclosure of errors
+
+## Type Safety of `errors` and `errorMap`
```

Comment:
In this paragraph I think it's important to explain that each key in the `errorMap` has exactly the type returned by its validator, while `errors` is an union type between all the errorMap values